clash-prelude/src/Clash/XException.hs

```diff
@@ -307,12 +306,12 @@ maybeIsX = maybeX isX
 -- > hasX _|_                 = _|_
 --
 -- If a data structure contains multiple 'XException's, the "first" message is
--- picked according to the implementation of 'rnf'.
-hasX :: NFData a => a -> Either String a
+-- picked according to the implementation of 'rnfX'.
+hasX :: NFDataX a => a -> Either String a
 hasX a =
   unsafeDupablePerformIO
     (catch
-      (evaluate (rnf a) >> return (Right a))
+      (evaluate (rnfX a) >> return (Right a))
```

Comment:
This works for things like: `hasX (3, errorX “QQ”)`? I’d be surprised, since rnfX would return the tuple, not throw the XException.

Reply:
Right... it should be `rnf (forceX ..)` I suppose. These would cause two traversals unfortunately.

Reply:
I think you need to do:
```
catch (evaluate (rnf a) >> return (Right a))
 (\(XException msg) -> evaluate (rnfX a) >> return (Left msg))
```
So you need both the NFData and NFDataX constraint.

Reply:
Alright, that makes sense. I'll add a TODO stating we could make `hasX` a type class method to prevent multiple traversals (even in the worst case).